pkg/reconciler/workload/synctargetexports/synctargetexports_reconcile.go

```diff
@@ -0,0 +1,167 @@
+}
+
+// exportReconciler updates syncedResource in SyncTarget status based on supporteAPIExports.
+type exportReconciler struct {
+	getAPIExport      func(clusterName logicalcluster.Name, name string) (*apisv1alpha1.APIExport, error)
+	getResourceSchema func(clusterName logicalcluster.Name, name string) (*apisv1alpha1.APIResourceSchema, error)
+}
+
+func (e *exportReconciler) reconcile(ctx context.Context, syncTarget *workloadv1alpha1.SyncTarget) (*workloadv1alpha1.SyncTarget, error) {
+	exportKeys := getExportKeys(syncTarget)
+
+	var errs []error
+	var syncedResources []workloadv1alpha1.ResourceToSync
+	for _, exportKey := range exportKeys {
+		exportCluster, name := clusters.SplitClusterAwareKey(exportKey)
+		export, err := e.getAPIExport(exportCluster, name)
+		if apierrors.IsNotFound(err) {
+			continue
+		}
+		if err != nil {
+			errs = append(errs, err)
+		}
+
+		for _, schema := range export.Spec.LatestResourceSchemas {
+			syncedResource, err := e.convertSchemaToSyncedResource(exportCluster, schema, export.Status.IdentityHash)
+			if err != nil {
+				klog.Warningf("cannot get schema: %v", err)
+				continue
+			}
+			syncedResources = append(syncedResources, syncedResource)
+		}
+	}
+
+	// sort synced resource by group
+	sort.SliceStable(syncedResources, func(i, j int) bool {
+		if syncedResources[i].Group > syncedResources[j].Group {
+			return true
+		}
+
+		if syncedResources[i].Resource > syncedResources[j].Resource {
+			return true
+		}
+		return false
+	})
+
+	// merge synced resource using desired as base and update it state based on existing synced.
+	for _, existingSynced := range syncTarget.Status.SyncedResources {
+		for i := range syncedResources {
+			if syncedResources[i].GroupResource == existingSynced.GroupResource && syncedResources[i].IdentityHash == existingSynced.IdentityHash {
+				syncedResources[i].State = existingSynced.State
+				break
+			}
+		}
+	}
+
+	syncTarget.Status.SyncedResources = syncedResources
+	return syncTarget, errors.NewAggregate(errs)
+}
+
+func (e *exportReconciler) convertSchemaToSyncedResource(cluterName logicalcluster.Name, schemaName, identityHash string) (workloadv1alpha1.ResourceToSync, error) {
+	schema, err := e.getResourceSchema(cluterName, schemaName)
+	if err != nil {
+		return workloadv1alpha1.ResourceToSync{}, err
+	}
+
+	syncedResource := workloadv1alpha1.ResourceToSync{
+		GroupResource: apisv1alpha1.GroupResource{
+			Group:    schema.Spec.Group,
+			Resource: schema.Spec.Names.Plural,
+		},
+		Versions:     []string{},
+		IdentityHash: identityHash,
+	}
+
+	for _, version := range schema.Spec.Versions {
+		if version.Served {
+			syncedResource.Versions = append(syncedResource.Versions, version.Name)
+		}
+	}
+	sort.Strings(syncedResource.Versions)
+
+	return syncedResource, nil
+}
+
+func (c *Controller) reconcile(ctx context.Context, syncTarget *workloadv1alpha1.SyncTarget) (*workloadv1alpha1.SyncTarget, error) {
```

Comment:
Why are these methods of `*Controller` not in the `synctargetexports_controller.go` file ?

Reply:
I agree on the general principle but still find it could seem surprising here, especially since this method is quite simple and just delegates to the 2 reconcilers, on or which is a distinct file.
I'm not opinionated though.